src/ast/mod.rs

```diff
@@ -51,11 +52,38 @@ pub(crate) enum Ruleset {
     Combined(Symbol, Vec<Symbol>),
 }
 
-pub type NCommand = GenericNCommand<Symbol, Symbol, ()>;
+pub(crate) const DEFAULT_FILENAME: &str = "<unnamed.egg>";
+pub(crate) const DUMMY_FILENAME: &str = "<internal.egg>";
+
+/// A [`Span`] contains the file name and a pair of offsets representing the start and the end.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Span(pub Symbol, pub usize, pub usize);
+
+lazy_static! {
+    pub(crate) static ref DUMMY_SPAN: Span = Span(Symbol::from(DUMMY_FILENAME), 0, 0);
```

Comment:
we talked about this in the meeting, but we should either not have a dummy span or make the annotation not generic

Reply:
+1; furthermore, I think having a single annotation type and not having type parameters may be the better option. Up to Yihong though.